Document ignore_invalidation_older_than

mkindahl opened the pull request:
Update the API documentation, tutorial, and user documentation to add
information about the `ignore_invalidation_older_than` parameter for
continous aggregates.

Suttask of: #622

gayyappan commented on api.md:
```diff
@@ -2253,23 +2253,25 @@ Get metadata and settings information for continuous aggregates.
 |`refresh_lag` | Amount by which the materialization for the continuous aggregate lags behind the current time |
 |`refresh_interval` | Interval between updates of the continuous aggregate materialization|
 |`max_interval_per_job` | Maximum amount of data processed by a materialization job in a single run|
+|`ignore_invalidation_older_than` | Do not update continous aggregate view for older updates|
```
Could you reword to use the same style as the docs for other parameters (passive vs active)?

gayyappan commented on using-timescaledb/continuous-aggregates.md:
```diff
@@ -151,6 +151,13 @@ will be spawned immediately following completion. This means that the part of th
+not trigger an update so, for example, setting the parameter to "1
+month" will ensure that any update for a record that is older than 1
+month will not trigger an update of the view.
+
```
we should also add a note here about how this affects data retention. Is that documented anywhere else? (maybe a link to drop_chunks_policy or material from a  blog).

gayyappan replied:
ok, lets do this when we add the section to data-retention.md. Need an example on how to do data retention for continuous aggregates as part of that section.

mkindahl replied:
I will add a section as a separate PR.

mkindahl replied:
It is available in https://github.com/timescale/docs.timescale.com-content/pull/314

solugebefola commented on api.md:
```diff
@@ -2253,23 +2253,25 @@ Get metadata and settings information for continuous aggregates.
+max_interval_per_job           | 20
+materialization_hypertable     | _timescaledb_internal._materialized_hypertable_2
+view_definition                |  SELECT foo.a,                                  +
+                               |     count(foo.b) AS countb                      +
```
Should 'count' be capitalized here?

solugebefola commented on tutorials/continuous-aggs-tutorial.md:
````diff
@@ -129,14 +129,20 @@ We can view the metadata for the continuous aggregate in the
 ``` sql
-SELECT view_name, refresh_lag, refresh_interval, max_interval_per_job, materialization_hypertable
+SELECT view_name
+     , refresh_lag
````
Is this a common syntax to have the comma preceding vs. comma at the end of the previous line?

mkindahl replied:
Both styles are common, I just pick this out of habit. We should have a common style in our documents and stick to it, so whatever is decided I will use.

gayyappan commented on api.md:
```diff
@@ -2253,23 +2253,25 @@ Get metadata and settings information for continuous aggregates.
+view_owner                     | postgres
+refresh_lag                    | 2
+refresh_interval               | 12:00:00
+ignore_invalidation_older_than | 1 month
```
ignore_invalidation_older_than looks incorrect.

mkindahl replied:
In what sense? It cannot be set to `1 month` so I changed that to `7 days` (and verified that it can actually be set).

gayyappan replied:
refresh_lag says 2, so thought this is a cont. aggregate on an integer based hypertable. Is it?

mkindahl replied:
Right, this is what I get when I run the tutorial:
```
nyc_data=# SELECT * FROM timescaledb_information.continuous_aggregates;
-[ RECORD 1 ]------------------+----------------------------------------------------------------------------------------
view_name                      | cagg_rides_view
view_owner                     | postgres
refresh_lag                    | 02:00:00
refresh_interval               | 00:30:00
max_interval_per_job           | 20:00:00
ignore_invalidation_older_than | 106751991 days 04:00:54.775807
materialization_hypertable     | _timescaledb_internal._materialized_hypertable_2
view_definition                |  SELECT rides.vendor_id,                                                               +
                               |     time_bucket('01:00:00'::interval, rides.pickup_datetime) AS day,                   +
                               |     count(*) AS total_rides,                                                           +
                               |     avg(rides.fare_amount) AS avg_fare,                                                +
                               |     max(rides.trip_distance) AS max_trip_distance,                                     +
                               |     min(rides.trip_distance) AS min_trip_distance                                      +
                               |    FROM rides                                                                          +
                               |   GROUP BY rides.vendor_id, (time_bucket('01:00:00'::interval, rides.pickup_datetime));
```

gayyappan commented on tutorials/continuous-aggs-tutorial.md:
````diff
@@ -99,7 +99,7 @@ continuous aggregate query every 30 minutes.
 ```sql
 CREATE VIEW cagg_rides_view WITH
-(timescaledb.continuous, timescaledb.refresh_interval = ’30m’)
+(timescaledb.continuous, timescaledb.refresh_interval = '30m')
````
Could you also add an example on how to create continuous aggregate views for hypertables that use "integer" based timestamps. Occasional question on support slack. (There is an example in tsl/test/sql/continuous_aggs.sql).  Either here or to the API docs,  whichever you think is more appropriate.

mkindahl replied:
I think it should be in the tutorial. I'll add it.

mkindahl replied:
From what I understand, integer-based timestamps are in microseconds. (Kind of fine-grained if you ask me. Easy to make a mistake and make the system unusable. Is there a request for microsecond `refresh_interval`?)

mkindahl replied:
Added just a paragraph about that in the tutorial.

gayyappan replied:
@mkindahl No, that's not the case. It can be anything the user wants to interpret it as, not necessarily microseconds. (The new paragraph should be removed.)  You will find an example with inetegr based timestamp and cont. aggregates here:  tsl/test/sql/continuous_aggs_bgw_drop_chunks.sql

mkindahl commented:
> Please label PR with earliest version

Not sure how. This is for the 1.7 release, so how should I label the PR?

gayyappan commented on api.md:
```diff
@@ -2253,23 +2253,25 @@ Get metadata and settings information for continuous aggregates.
 |`refresh_lag` | Amount by which the materialization for the continuous aggregate lags behind the current time |
 |`refresh_interval` | Interval between updates of the continuous aggregate materialization|
 |`max_interval_per_job` | Maximum amount of data processed by a materialization job in a single run|
+|`ignore_invalidation_older_than` | Interval for which updated rows will trigger update of the continuous aggregate|
```
typo -> will not
```suggestion
|`ignore_invalidation_older_than` | Interval for which modified rows (i.e. insert/update/delete) will not trigger update of the continuous aggregate|
```

mkindahl replied:
Shouldn't it be "will" since updates in the interval [NOW - IIOT, NOW] will trigger an update, or have I misunderstood how it works?

gayyappan replied:
I think of it as ignoring anything before NOW - IIOT. Maybe we should not refer to it as _interval_ as it is not clear which interval we are referring to : is it anything < NOW - IIOT or the other one.

mkindahl replied:
Yes, but not sure how to re-write it without using active tense. "Age before which modifications do not trigger an update of the continuous aggregate."?

gayyappan replied:
lets do that and see if anyone else has suggestions.

gayyappan commented on using-timescaledb/continuous-aggregates.md:
```diff
@@ -151,6 +151,13 @@ will be spawned immediately following completion. This means that the part of th
 data processed by previous jobs will be available even before the
 materialization is fully caught up.
 
+The `timescaledb.ignore_invalidation_older_than` parameter control how
```
```suggestion
The `timescaledb.ignore_invalidation_older_than` parameter controls how
```

mkindahl replied:
Done.

gayyappan commented on using-timescaledb/continuous-aggregates.md:
```diff
@@ -151,6 +151,13 @@ will be spawned immediately following completion. This means that the part of th
+not trigger an update so, for example, setting the parameter to "1
+month" will ensure that any update for a record that is older than 1
+month will not trigger an update of the view.
+
```
```suggestion
By default, all invalidations are processed. When ignore_invalidation_older_than is specified, the view is freed from the underlying data, thus allowing those data points to exist as they are in the continuous aggregate without worrying about changes to the data invalidating the view. As a result, refresh jobs will complete faster since they are only dealing with new data as part of the refresh, and you will also have the option to implement a data retention policy. <link to data retention policy section>
```
(adapted from blog material).

mkindahl replied:
This is incredibly hard to read:

> By default, all invalidations are processed. When `ignore_invalidation_older_than` is specified, the view is freed from the underlying data, thus allowing those data points to exist as they are in the continuous aggregate without worrying about changes to the data invalidating the view.

What does it mean to be "freed from the underlying data"? What does it mean for a data point to "exist as they are"? It is not clear what an "invalidation" is, it is not described earlier in the text.

Going to try with some other rewrite.

mkindahl replied:
For the second part:

> As a result, refresh jobs will complete faster since they are only dealing with new data as part of the refresh, and you will also have the option to implement a data retention policy. <link to data retention policy section>

What does it mean to have the "option to implement a data retention policy"? Can you not use that if you do not set a value for `ignore_invalidation_older_than`, or is it just that it does not make sense?

mkindahl replied:
Rewrote to something. PTAL.

mkindahl replied:
Rewritten and some clarifications in other threads. Resolving.

gayyappan commented on using-timescaledb/continuous-aggregates.md:
```diff
@@ -151,6 +151,13 @@ will be spawned immediately following completion. This means that the part of th
 data processed by previous jobs will be available even before the
 materialization is fully caught up.
 
+The `timescaledb.ignore_invalidation_older_than` parameter control how
```
Should we add a new subsection here:  "More ways to control continuous aggregate?" or something similar. Then add ignore_invalidation_older_than and the parameter from Sven's JIT aggregates in this scetion.

mkindahl replied:
We probably need an additional section for JIT, yes, but I should add that as a separate PR.

gayyappan commented on using-timescaledb/continuous-aggregates.md:
```diff
@@ -151,6 +151,25 @@ will be spawned immediately following completion. This means that the part of th
+By default, all modifications trigger a recompute of the continuous
+aggregate. If the `ignore_invalidation_older_than` parameter is set to
+a duration, modifications with a timestamp that is older will be
+ignored and not trigger a recompute of the continuous aggregate.
```
Like the rewrite here.
```suggestion
ignored and not trigger an update of the continuous aggregate.
```

gayyappan commented on using-timescaledb/continuous-aggregates.md:
```diff
@@ -151,6 +151,25 @@ will be spawned immediately following completion. This means that the part of th
+As a result, refresh jobs will complete faster since they are only
+dealing with new data as part of the refresh.
+
+If setting `ignore_invalidation_older_than` you need to consider if
```
Data retention policies for hyperatbles are affected by continuous aggregates. Say you have a raw table and a cont agg defned on it. If the hypertable has drop chunks wants to drop chunks older than 1 month, then the ignore_invalidation-older_than parameter for cont. aggregates that are dependent on it should be 1 month. So dropping the raw data does not affect the continuous aggregate. (I think the detailed explanation can wait for the data retention policy PR. But adding a link here to data retention doc is probably useful).

mkindahl replied:
This sounds dangerous, should we not disallow dropping chunks if it can affect a continuous aggregate? Need to have a good error message as well.

gayyappan replied:
We check if there are cont. aggregates dependent on the chunk before it can be dropped. Errors, otherwise.

mkindahl replied:
Maybe that should be mentioned in the documentation?

mkindahl replied:
I'll add this to the data retention PR.

gayyappan commented on using-timescaledb/continuous-aggregates.md:
```diff
@@ -151,6 +151,25 @@ will be spawned immediately following completion. This means that the part of th
+dealing with new data as part of the refresh.
+
+If setting `ignore_invalidation_older_than` you need to consider if
+the data retention policy need to be updated. For more information on
```
So I would rephrase this as (or  a variation).
```suggestion
Often users would like to drop the raw data from the hypertable and just keep the downsampled data in the continuous aggregate. You can do this by using the invalidation-older_than setting of the continuous aggregate. For e.g. by setting the invalidation_older_than parameter to 1 month, we are saying that continuous aggregate will not be modified for any changes older than 1 month. So a user can drop_chunks from the hypertable that is older than 1 month and still retain the downsampled data in the continuous aggregate. You cna read more about data retention <here>.
```

gayyappan commented on using-timescaledb/continuous-aggregates.md:
```diff
@@ -151,6 +151,25 @@ will be spawned immediately following completion. This means that the part of th
+The `timescaledb.ignore_invalidation_older_than` parameter controls
+how modifications (inserts, updates, and deletes) will trigger update
+of the continuous aggregate. If modifications are done to the
+underlying table (or tables), it invalidates the already computed
```
```suggestion
underlying table, it invalidates the already computed
```
cont. aggregate can be defined only on a single hypertable.

gayyappan commented on using-timescaledb/continuous-aggregates.md:
```diff
@@ -151,6 +151,25 @@ will be spawned immediately following completion. This means that the part of th
+how modifications (inserts, updates, and deletes) will trigger update
+of the continuous aggregate. If modifications are done to the
+underlying table (or tables), it invalidates the already computed
+aggregate and the aggregate has to be recomputed.
```
```suggestion
aggregate and the aggregate has to be updated.
```
we only have to update the invalidated portion.

mkindahl commented:
> > Please label PR with earliest version
> 
> Not sure how. This is for the 1.7 release, so how should I label the PR?

Added a new milestone "1.7" and added the PR to it.

gayyappan commented on api.md:
```diff
@@ -2253,23 +2253,25 @@ Get metadata and settings information for continuous aggregates.
+view_definition                |  SELECT foo.a,                                  +
+                               |     COUNT(foo.b) AS countb                      +
+                               |    FROM foo                                     +
+                               |   GROUP BY (time_bucket(1, foo.a)), foo.a;
```
I think the view stmt is for a integer based hypertable
changing the time_bucket to time_bucket('1 day', ...) should fix that.

mkindahl replied:
Done.

gayyappan commented on tutorials/continuous-aggs-tutorial.md:
```diff
@@ -168,6 +170,13 @@ smaller or bigger batch sizes (the batching is done automatically by TimescaleDB
 
+The `timescaledb.ignore_invalidation_older_than` parameter is used
+when you want to avoid updating a continuous aggregate when modifying
+(that is, inserting, updating, or deleting) older record. Setting this
```
```suggestion
(that is, inserting, updating, or deleting) older records. Setting this
```
